### PR TITLE
feature: add LZ4 compressor

### PR DESCRIPTION
**Pull request by diguage**
LZ4 is an extremely fast Compression algorithm. Maybe it should be supported.

* [lz4/lz4: Extremely Fast Compression algorithm](https://github.com/lz4/lz4#benchmarks)
* [RealTime Data Compression: LZ4 explained](http://fastcompression.blogspot.com/2011/05/lz4-explained.html)

### CODE REVIEW

**Comment by zjinlei** [comment]
please fix CI error

**Comment by a364176773** [comment]
please format the code according to the p3c specification

**Comment by diguage** [comment]
> please fix CI error:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project seata-compressor-lz4: Failed during checkstyle execution: There are 4 errors reported by Checkstyle 6.18 with /home/travis/build/seata/seata/style/seata_checkstyle.xml ruleset.

There is an error. And there is no report. So I cannot find it.

@a364176773 I use p3c to check the code. There is not any warning.

Please help me.

**Comment by a364176773** [comment]
> > please fix CI error:
> > [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project seata-compressor-lz4: Failed during checkstyle execution: There are 4 errors reported by Checkstyle 6.18 with /home/travis/build/seata/seata/style/seata_checkstyle.xml ruleset.
> 
> There is an error. And there is no report. So I cannot find it.
> 
> @a364176773 I use p3c to check the code. There is not any warning.
> 
> Please help me.

please check the code format and local mvn clean install -DskipTests=true

**Comment by diguage** [comment]
> please check the code format and local mvn clean install -DskipTests=true

I ran the command. The log had an error. But did not show the detail.

The log is as following:

```
...
[INFO] seata-spring-boot-starter 1.2.0-SNAPSHOT ........... SUCCESS [  1.714 s]
[INFO] seata-compressor-lz4 1.2.0-SNAPSHOT ................ FAILURE [  0.235 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:27 min
[INFO] Finished at: 2020-04-07T12:39:15+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project seata-compressor-lz4: Failed during checkstyle execution: There is 1 error reported by Checkstyle 6.18 with /Users/diguage/develop/java/javaprojects/seata/style/seata_checkstyle.xml ruleset. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :seata-compressor-lz4
```

**Comment by a364176773** [comment]
> > please check the code format and local mvn clean install -DskipTests=true
> 
> I ran the command. The log had an error. But did not show the detail.
> 
> The log is as following:
> 
> ```
> ...
> [INFO] seata-spring-boot-starter 1.2.0-SNAPSHOT ........... SUCCESS [  1.714 s]
> [INFO] seata-compressor-lz4 1.2.0-SNAPSHOT ................ FAILURE [  0.235 s]
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  01:27 min
> [INFO] Finished at: 2020-04-07T12:39:15+08:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project seata-compressor-lz4: Failed during checkstyle execution: There is 1 error reported by Checkstyle 6.18 with /Users/diguage/develop/java/javaprojects/seata/style/seata_checkstyle.xml ruleset. -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> [ERROR]
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <goals> -rf :seata-compressor-lz4
> ```

scroll up to find the corresponding module log, which should have a prompt message

**Comment by diguage** [comment]
> > > please check the code format and local mvn clean install -DskipTests=true
> > 
> > 
> > I ran the command. The log had an error. But did not show the detail.
> > The log is as following:
> > ```
> > ...
> > [INFO] seata-spring-boot-starter 1.2.0-SNAPSHOT ........... SUCCESS [  1.714 s]
> > [INFO] seata-compressor-lz4 1.2.0-SNAPSHOT ................ FAILURE [  0.235 s]
> > [INFO] ------------------------------------------------------------------------
> > [INFO] BUILD FAILURE
> > [INFO] ------------------------------------------------------------------------
> > [INFO] Total time:  01:27 min
> > [INFO] Finished at: 2020-04-07T12:39:15+08:00
> > [INFO] ------------------------------------------------------------------------
> > [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project seata-compressor-lz4: Failed during checkstyle execution: There is 1 error reported by Checkstyle 6.18 with /Users/diguage/develop/java/javaprojects/seata/style/seata_checkstyle.xml ruleset. -> [Help 1]
> > [ERROR]
> > [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> > [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> > [ERROR]
> > [ERROR] For more information about the errors and possible solutions, please read the following articles:
> > [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> > [ERROR]
> > [ERROR] After correcting the problems, you can resume the build with the command
> > [ERROR]   mvn <goals> -rf :seata-compressor-lz4
> > ```
> 
> scroll up to find the corresponding module log, which should have a prompt message

I saw the `seata/compressor/seata-compressor-lz4/target/checkstyle-result.xml` file. There is a tip.

Thank you very much.

**Comment by slievrly** [comment]
@diguage mvn clean install -DskipTests=true -Dmaven.javadoc.skip=true -Dcheckstyle.skip=false

**Comment by diguage** [comment]
> @diguage mvn clean install -DskipTests=true -Dmaven.javadoc.skip=true -Dcheckstyle.skip=false

Thank you. I found the error, and fixed it. Thanks again.

File: compressor/seata-compressor-lz4/pom.xml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>seata-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.lz4</groupId>
+            <artifactId>lz4-java</artifactId>
+            <version>1.7.1</version>
```
Comment (ph3636) [inline compressor/seata-compressor-lz4/pom.xml:39]: Please put version into all and bom files. For details, please refer to bzip2

File: compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import net.jpountz.lz4.LZ4BlockOutputStream;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4FastDecompressor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
```
Comment (ph3636) [inline compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:28]: Java package at the top,Change BZip2Util and SevenZUtil by the way, if you can

File: compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        return outputStream.toByteArray();
+    }
+
+    public static byte[] decompress(byte[] bytes) {
+        if (bytes == null) {
+            throw new NullPointerException("bytes is null");
+        }
+
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream(ARRAY_SIZE);
```
Comment (ph3636) [inline compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:61]: You don't need to initialize such a large value, just keep the default value

File: compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * the Lz4 Util
+ *
+ * @author diguage
+ */
+public class Lz4Util {
+    private static final Logger LOGGER = LoggerFactory.getLogger(Lz4Util.class);
+    private static final int ARRAY_SIZE = 2048;
+
+    private static final LZ4Factory LZ4_FACTORY = LZ4Factory.fastestInstance();
```
Comment (ph3636) [inline compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:39]: This one can get rid of static decoration

File: compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java (hunk 3 above)
Comment (diguage) [reply compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:61]: The default size is 32.  Is it too small?

File: compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java (hunk 4 above)
Comment (diguage) [reply compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:39]: The `LZ4_FACTORY` is used in the `static` methods.

Comment (ph3636) [reply compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:39]: Remove the static variable and use it directly in the method

Comment (ph3636) [reply compressor/seata-compressor-lz4/src/main/java/io/seata/compressor/lz4/Lz4Util.java:39]: Remove the static variable and use it directly in the method

File: compressor/seata-compressor-lz4/pom.xml (hunk 1 above)
Comment (ph3636) [reply compressor/seata-compressor-lz4/pom.xml]: seata-all file also needs to contain seata-compressor-lz4

**Comment by ph3636** [comment]
seata-all file is missing some configuration

**Comment by diguage** [comment]
> seata-all file is missing some configuration

OK

**Comment by zjinlei** [comment]
Please keep the commit log, we can only review the increment.

**Comment by ph3636** [comment]
> > seata-all file is missing some configuration
> 
> OK

The 682 line of seata-al file is missing some configuration

**Comment by diguage** [comment]
> The 682 line of seata-all file is missing some configuration

OK 👌

**Comment by diguage** [comment]
> Please keep the commit log, we can only review the increment.
👌 Thanks.

File: all/pom.xml [hunk 5]
```diff
@@ -284,6 +284,11 @@
             <artifactId>seata-compressor-zip</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
```
Comment (slievrly) [inline all/pom.xml:287]: pls add `lz4-java` dependency in all.pom

Comment (diguage) [reply all/pom.xml:287]: OK👌
Please review.

**Pull request merged**